File: src/lib/parser/parser_test.ts
import { describe, it } from "jsr:@std/testing/bdd"
import { join } from "jsr:@std/path"
import { expect } from "jsr:@std/expect"
import { parseBlocks } from "./parser.ts"
import { findBlocks } from "./blocks.ts"

const __dirname = new URL(".", import.meta.url).pathname
const exampleLayoutFilename = join(__dirname, "..", "..", "..", "examples", "iris.md")
const exampleLayout = Deno.readTextFileSync(exampleLayoutFilename)

describe(parseBlocks.name, () => {
  const blocks = findBlocks(exampleLayout.split("\n"))
  const layout = parseBlocks(blocks)

  it("parses the options block", () => {
    expect(layout.options).toEqual({
      layoutFn: "LAYOUT",
    })
  })

  it("parses the aliases block", () => {
    const aliases = Array.from(layout.aliases!.entries())
    expect(aliases).toEqual([["lock", { value: "c+g+q", lineNr: 12 }]])
  })

  it("parses the combos block", () => {
    expect(layout.combos).toEqual([{ keys: "q+w", action: "esc", lineNr: 16 }])
  })
})

File: internal/parser/parser_test.go
package parser

import (
	"path/filepath"
	"testing"
)

func TestParseBlocksFromExample(t *testing.T) {
	t.Parallel()

	filename := filepath.Join("..", "..", "examples", "iris.md")
	_, blocks, err := ParseMarkdownFile(filename)
	if err != nil {
		t.Fatalf("ParseMarkdownFile returned error: %v", err)
	}

	parsed, err := ParseBlocks(blocks)
	if err != nil {
		t.Fatalf("ParseBlocks returned error: %v", err)
	}

	if parsed.Options.LayoutFn != "LAYOUT" {
		t.Fatalf("expected layout fn LAYOUT, got %q", parsed.Options.LayoutFn)
	}
	if len(parsed.Aliases) == 0 {
		t.Fatalf("expected aliases to be parsed")
	}
	lock, ok := parsed.Aliases["lock"]
	if !ok {
		t.Fatalf("expected lock alias to exist")
	}
	if lock.Value != "c+g+q" || lock.LineNr != 12 {
		t.Fatalf("unexpected lock alias: %+v", lock)
	}
	if len(parsed.Combos) == 0 {
		t.Fatalf("expected combos to be parsed")
	}
	if parsed.Combos[0].Keys != "x+c" || parsed.Combos[0].Action != "lalt" || parsed.Combos[0].LineNr != 22 {
		t.Fatalf("unexpected first combo: %+v", parsed.Combos[0])
	}
	if len(parsed.Layers) == 0 {
		t.Fatalf("expected layers to be parsed")
	}
	if parsed.Layers[0].Name != "base" {
		t.Fatalf("expected first layer to be base, got %q", parsed.Layers[0].Name)
	}
}
